java-frontend/src/main/java/org/sonar/java/collections/AVLTree.java

```diff
@@ -263,63 +273,105 @@ private static AVLTree removeMinBinding(AVLTree t, NodeRef noderemoved) {
+  private static AVLTree balance(AVLTree l, AVLTree oldNode, AVLTree r) {
     if (l.height() > r.height() + 2) {
       assert !l.isEmpty();
       AVLTree ll = l.left();
       AVLTree lr = l.right();
       if (ll.height() >= lr.height()) {
-        return createNode(ll, l, createNode(lr, key, value, r));
+        return createNode(ll, l, createNode(lr, oldNode, r));
       }
       assert !lr.isEmpty();
       AVLTree lrl = lr.left();
       AVLTree lrr = lr.right();
-      return createNode(createNode(ll, l, lrl), lr, createNode(lrr, key, value, r));
+      return createNode(createNode(ll, l, lrl), lr, createNode(lrr, oldNode, r));
     }
     if (r.height() > l.height() + 2) {
       assert !r.isEmpty();
       AVLTree rl = r.left();
       AVLTree rr = r.right();
       if (rr.height() >= rl.height()) {
-        return createNode(createNode(l, key, value, rl), r, rr);
+        return createNode(createNode(l, oldNode, rl), r, rr);
       }
       assert !rl.isEmpty();
       AVLTree rll = rl.left();
       AVLTree rlr = rl.right();
-      return createNode(createNode(l, key, value, rll), rl, createNode(rlr, r, rr));
+      return createNode(createNode(l, oldNode, rll), rl, createNode(rlr, r, rr));
     }
-    return createNode(l, key, value, r);
+    return createNode(l, oldNode, r);
   }
 
   private static AVLTree createNode(AVLTree newLeft, AVLTree oldTree, AVLTree newRight) {
-    return createNode(newLeft, oldTree.key(), oldTree.value(), newRight);
+    return new Node(newLeft, newRight, oldTree.key(), oldTree.value(), oldTree.nextInBucket(), incrementHeight(newLeft, newRight));
   }
 
-  private static AVLTree createNode(AVLTree l, Object key, Object value, AVLTree r) {
-    return new Node(l, r, key, value, incrementHeight(l, r));
+  private static Node createNode(AVLTree l, Object key, Object value, @Nullable AVLTree nextInBucket, AVLTree r) {
+    return new Node(l, r, key, value, nextInBucket, incrementHeight(l, r));
   }
 
   private static int incrementHeight(AVLTree l, AVLTree r) {
     return (l.height() > r.height() ? l.height() : r.height()) + 1;
   }
 
-  private static class Node extends AVLTree {
+  /**
+   * @return node from the given bucket, which contains given key, null if not found
+   */
+  @Nullable
+  private static AVLTree searchInBucket(Object key, @Nullable AVLTree bucketStart) {
+    AVLTree n = bucketStart;
+    while (n != null) {
+      if (key.equals(n.key())) {
+        return n;
+      }
+      n = n.nextInBucket();
+    }
+    return null;
+  }
+
+  /**
+   * Given "b" as a node to remove and bucket "a -> b -> c" result will be "c -> a".
+   *
+   * @return null if bucket is null
+   */
+  @Nullable
+  private static AVLTree removeFromBucket(@Nullable AVLTree bucketStart, @Nullable AVLTree nodeToRemove) {
+    AVLTree c = bucketStart;
+    AVLTree result = null;
+    while (c != null) {
+      if (/* not the instance to remove: */ c != nodeToRemove) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1698)